In [154]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [155]:
df2019 = pd.read_csv('/home/pato/Documents/GitHub/tarea1-add/data/interim/d2019.csv') 
df2020 = pd.read_table('/home/pato/Documents/GitHub/tarea1-add/data/interim/d2020.txt', sep= ' ')

In [177]:
df2019.head()

Día     Ene     Feb     Mar     Abr     May     Jun     Jul     Ago  \
0    1     NaN  657.81  651.79  678.53     NaN     NaN  679.15  700.82   
1    2  694.77     NaN     NaN  672.56  678.71     NaN  677.62  704.88   
2    3  697.09     NaN     NaN  670.05  683.31  709.80  681.30     NaN   
3    4  697.64  655.39  656.57  667.55     NaN  704.59  680.48     NaN   
4    5     NaN  653.80  659.07  666.32     NaN  696.78  677.64  711.13   

      Sep     Oct     Nov     Dic    year  
0     NaN  728.21     NaN     NaN  2019.0  
1  720.20  729.38     NaN  812.13  2019.0  
2  724.72  727.80     NaN  804.29  2019.0  
3  725.69  719.53  735.05  802.44  2019.0  
4  723.30     NaN  736.66  796.75  2019.0

In [176]:
df2020.head()

Día     Ene     Feb     Mar     Abr     May     Jun     Jul     Ago  \
0    1     NaN     NaN     NaN  852.03     NaN  806.32  821.23     NaN   
1    2  748.74     NaN  818.32  863.81     NaN  796.46  816.29     NaN   
2    3  754.16  799.11  817.45  861.89     NaN  782.86  803.98  757.06   
3    4     NaN  790.92  809.81     NaN  837.92  769.13     NaN  761.34   
4    5     NaN  781.78  812.19     NaN  838.74  775.58     NaN  767.98   

      Sep     Oct     Nov     Dic    year  
0  776.46  788.15     NaN  767.29  2020.0  
1  770.59  785.04  771.92  760.16  2020.0  
2  771.45     NaN  769.17  755.34  2020.0  
3  773.71     NaN  758.53  752.03  2020.0  
4     NaN  790.81  757.16     NaN  2020.0

In [158]:
df_sii = pd.melt(pd.concat([df2019,df2020], axis=0), id_vars=['year', 'Día'])

In [159]:
df_sii = df_sii.rename(columns={'variable':'mes', 'value':'Valor'})
df_sii

year  Día  mes   Valor
0    2019.0    1  Ene     NaN
1    2019.0    2  Ene  694.77
2    2019.0    3  Ene  697.09
3    2019.0    4  Ene  697.64
4    2019.0    5  Ene     NaN
..      ...  ...  ...     ...
739  2020.0   27  Dic     NaN
740  2020.0   28  Dic  710.26
741  2020.0   29  Dic  710.64
742  2020.0   30  Dic  711.24
743  2020.0   31  Dic     NaN

[744 rows x 4 columns]

In [160]:
df_sii.to_csv('/home/pato/Documents/GitHub/tarea1-add/data/interim/df_sii.csv')

In [161]:
df_sii.groupby(['mes'], as_index=True).aggregate('mean')

year   Día       Valor
mes                          
Abr  2019.5  16.0  760.389048
Ago  2019.5  16.0  749.182619
Dic  2019.5  16.0  752.561750
Ene  2019.5  16.0  724.854773
Feb  2019.5  16.0  726.342000
Jul  2019.5  16.0  735.394318
Jun  2019.5  16.0  744.299024
Mar  2019.5  16.0  755.523488
May  2019.5  16.0  753.659500
Nov  2019.5  16.0  769.540976
Oct  2019.5  16.0  753.867907
Sep  2019.5  16.0  748.036154

In [162]:
df_sii.groupby(['year', 'mes'], as_index=False).aggregate('mean')

year  mes   Día       Valor
0   2019.0  Abr  16.0  667.399048
1   2019.0  Ago  16.0  713.703333
2   2019.0  Dic  16.0  770.390500
3   2019.0  Ene  16.0  677.061818
4   2019.0  Feb  16.0  656.304500
5   2019.0  Jul  16.0  686.059545
6   2019.0  Jun  16.0  692.409000
7   2019.0  Mar  16.0  667.678571
8   2019.0  May  16.0  692.003810
9   2019.0  Nov  16.0  776.530000
10  2019.0  Oct  16.0  721.032273
11  2019.0  Sep  16.0  718.442222
12  2020.0  Abr  16.0  853.379048
13  2020.0  Ago  16.0  784.661905
14  2020.0  Dic  16.0  734.733000
15  2020.0  Ene  16.0  772.647727
16  2020.0  Feb  16.0  796.379500
17  2020.0  Jul  16.0  784.729091
18  2020.0  Jun  16.0  793.718095
19  2020.0  Mar  16.0  839.375455
20  2020.0  May  16.0  821.805263
21  2020.0  Nov  16.0  762.884762
22  2020.0  Oct  16.0  788.267143
23  2020.0  Sep  16.0  773.402381

as_index tomara el valor de la primera columna y lo pondra como indice si el valor es true. Al cambiar de (mes) a (year, mes), el primero ordena por mes, pero toma de promedio el año y los dias, mientras que el ultimo ahora ordena por mes y por año.

In [163]:
#Eliminación de fechas inexistentes en calendario
#Febrero
df_sii=df_sii.drop(df_sii[(df_sii["mes"]=="Feb")&(df_sii["Día"]>=29)&(df_sii["year"]==2019)].index)
df_sii=df_sii.drop(df_sii[(df_sii["mes"]=="Feb")&(df_sii["Día"]>=30)&(df_sii["year"]==2020)].index) #año bisiesto
#Abril
df_sii=df_sii.drop(df_sii[(df_sii["mes"]=="Abr")&(df_sii["Día"]>=31)&(df_sii["year"]==2019)].index)
df_sii=df_sii.drop(df_sii[(df_sii["mes"]=="Abr")&(df_sii["Día"]>=31)&(df_sii["year"]==2020)].index)
#Junio
df_sii=df_sii.drop(df_sii[(df_sii["mes"]=="Jun")&(df_sii["Día"]>=31)&(df_sii["year"]==2019)].index)
df_sii=df_sii.drop(df_sii[(df_sii["mes"]=="Jun")&(df_sii["Día"]>=31)&(df_sii["year"]==2020)].index)
#Septiembre
df_sii=df_sii.drop(df_sii[(df_sii["mes"]=="Sep")&(df_sii["Día"]>=31)&(df_sii["year"]==2019)].index)
df_sii=df_sii.drop(df_sii[(df_sii["mes"]=="Sep")&(df_sii["Día"]>=31)&(df_sii["year"]==2020)].index)
#Noviembre
df_sii=df_sii.drop(df_sii[(df_sii["mes"]=="Nov")&(df_sii["Día"]>=31)&(df_sii["year"]==2019)].index)
df_sii=df_sii.drop(df_sii[(df_sii["mes"]=="Nov")&(df_sii["Día"]>=31)&(df_sii["year"]==2020)].index)

In [164]:
#creacion campos month y day

df_sii["month"] = np.where(df_sii['mes']=="Ene", 1,
                   np.where(df_sii['mes']=="Feb", 2,
                   np.where(df_sii['mes']=="Mar", 3,
                   np.where(df_sii['mes']=="Abr", 4,
                   np.where(df_sii['mes']=="May", 5,
                    np.where(df_sii['mes']=="Jun", 6,
                    np.where(df_sii['mes']=="Jul", 7,
                    np.where(df_sii['mes']=="Ago", 8,
                    np.where(df_sii['mes']=="Sep", 9,
                    np.where(df_sii['mes']=="Oct", 10,
                    np.where(df_sii['mes']=="Nov", 11,12)))))))))))    

df_sii["day"]=df_sii["Día"]

In [165]:
#creacion campo fechas
df_sii["fecha"]=pd.to_datetime(df_sii[['year', 'month', 'day']])

In [166]:
#creacion campo dia_de_la_semana
df_sii["dia_de_la_semana"]=df_sii['fecha'].dt.day_name()
df_sii["dia_de_la_semana"]=np.where(df_sii["dia_de_la_semana"]=="Monday","Lunes",
                           np.where(df_sii["dia_de_la_semana"]=="Tuesday","Martes",
                           np.where(df_sii["dia_de_la_semana"]=="Wednesday","Miercoles",
                           np.where(df_sii["dia_de_la_semana"]=="Thursday","Jueves",
                           np.where(df_sii["dia_de_la_semana"]=="Friday","Viernes",
                           np.where(df_sii["dia_de_la_semana"]=="Saturday","Sabado","Domingo"))))))

df_sii=df_sii[["year","mes","Día","Valor","dia_de_la_semana"]]
df_sii

year  mes  Día   Valor dia_de_la_semana
0    2019.0  Ene    1     NaN           Martes
1    2019.0  Ene    2  694.77        Miercoles
2    2019.0  Ene    3  697.09           Jueves
3    2019.0  Ene    4  697.64          Viernes
4    2019.0  Ene    5     NaN           Sabado
..      ...  ...  ...     ...              ...
739  2020.0  Dic   27     NaN          Domingo
740  2020.0  Dic   28  710.26            Lunes
741  2020.0  Dic   29  710.64           Martes
742  2020.0  Dic   30  711.24        Miercoles
743  2020.0  Dic   31     NaN           Jueves

[731 rows x 5 columns]

### No se ve que exista un cambio entre los dias de semana y fin de semana, comparandolos por año desde el 2019 hasta el final del 2020

In [167]:
#creación de un df para cada hoja del archivo excel
dolar_observado_bc = pd.read_excel("https://raw.githubusercontent.com/hfas85/tarea1-add/master/data/raw/dolar_observado_bc.xlsx",
                                 skiprows=2, dtype={'Periodo':str }, sheet_name=None)

df_1980 = dolar_observado_bc.get('80')
df_1990 = dolar_observado_bc.get('90')
df_2000 = dolar_observado_bc.get('00')
df_2010 = dolar_observado_bc.get('10')
df_2020 = dolar_observado_bc.get('20')

df_1980["Periodo"]=df_1980["Periodo"].str[0:10]
df_1990["Periodo"]=df_1990["Periodo"].str[0:10]
df_2000["Periodo"]=df_2000["Periodo"].str[0:10]
df_2010["Periodo"]=df_2010["Periodo"].str[0:10]
df_2020["Periodo"]=df_2020["Periodo"].str[0:10]

In [168]:
#creación de campos year,mes,dia para cada df
df_1980[["year", "mes", "dia"]] = df_1980["Periodo"].str.split("-", expand = True)
df_1990[["year", "mes", "dia"]] = df_1990["Periodo"].str.split("-", expand = True)
df_2000[["year", "mes", "dia"]] = df_2000["Periodo"].str.split("-", expand = True)
df_2010[["year", "mes", "dia"]] = df_2010["Periodo"].str.split("-", expand = True)
df_2020[["year", "mes", "dia"]] = df_2020["Periodo"].str.split("-", expand = True)

df_1980.columns = ["Periodo","Dolar","year","mes","dia"]
df_1990.columns = ["Periodo","Dolar","year","mes","dia"]
df_2000.columns = ["Periodo","Dolar","year","mes","dia"]
df_2010.columns = ["Periodo","Dolar","year","mes","dia"]
df_2020.columns = ["Periodo","Dolar","year","mes","dia"]

df_2020.head()

Periodo   Dolar  year mes dia
0  2020-01-01     NaN  2020  01  01
1  2020-01-02  748.74  2020  01  02
2  2020-01-03  754.16  2020  01  03
3  2020-01-04     NaN  2020  01  04
4  2020-01-05     NaN  2020  01  05

In [169]:
#creacion de un solo data frame
df_bc_all=pd.concat((df_1980,df_1990,df_2000,df_2010,df_2020),axis=0)
df_bc_all

Periodo   Dolar  year mes dia
0    1982-08-09   55.65  1982  08  09
1    1982-08-10   62.79  1982  08  10
2    1982-08-11   64.07  1982  08  11
3    1982-08-12   60.52  1982  08  12
4    1982-08-13   56.87  1982  08  13
..          ...     ...   ...  ..  ..
969  2022-08-27     NaN  2022  08  27
970  2022-08-28     NaN  2022  08  28
971  2022-08-29  900.22  2022  08  29
972  2022-08-30  890.90  2022  08  30
973  2022-08-31  882.11  2022  08  31

[14633 rows x 5 columns]

In [170]:
# Creación de csv agrupando por "year"
for (year), group in df_bc_all.groupby(['year']):
   group.to_csv(f'{year}.csv', index=False)

In [171]:
#importar csv de los años 2019 y 2020
d_2019_2=pd.read_csv("2019.csv")
d_2020_2=pd.read_csv("2020.csv")

df_bc_2019_2020=pd.concat((d_2019_2,d_2020_2),axis=0)

In [172]:
#comparación df_bc_2019_2020 con df_sii.
df_bc_2019_2020.head(10)

Periodo   Dolar  year  mes  dia
0  2019-01-01     NaN  2019    1    1
1  2019-01-02  694.77  2019    1    2
2  2019-01-03  697.09  2019    1    3
3  2019-01-04  697.64  2019    1    4
4  2019-01-05     NaN  2019    1    5
5  2019-01-06     NaN  2019    1    6
6  2019-01-07  688.42  2019    1    7
7  2019-01-08  679.69  2019    1    8
8  2019-01-09  680.10  2019    1    9
9  2019-01-10  678.09  2019    1   10

In [173]:
#comparación df_bc_2019_2020 con df_sii.
df_sii.head(10)

year  mes  Día   Valor dia_de_la_semana
0  2019.0  Ene    1     NaN           Martes
1  2019.0  Ene    2  694.77        Miercoles
2  2019.0  Ene    3  697.09           Jueves
3  2019.0  Ene    4  697.64          Viernes
4  2019.0  Ene    5     NaN           Sabado
5  2019.0  Ene    6     NaN          Domingo
6  2019.0  Ene    7  688.42            Lunes
7  2019.0  Ene    8  679.69           Martes
8  2019.0  Ene    9  680.10        Miercoles
9  2019.0  Ene   10  678.09           Jueves

In [174]:
#comparación df_bc_2019_2020 con df_sii.
df_bc_2019_2020.shape, df_sii.shape

((731, 5), (731, 5))

Los valores de los DF son iguales, la cantidad de registros en ambos es de 731

In [181]:
pd.merge(df_bc_all, df_sii.set_index('dia_de_la_semana'),
         left_on='dia',
         right_index=True,
         how='left')

Periodo   Dolar year_x mes_x dia  year_y mes_y  Día  Valor
0    1982-08-09   55.65   1982    08  09     NaN   NaN  NaN    NaN
1    1982-08-10   62.79   1982    08  10     NaN   NaN  NaN    NaN
2    1982-08-11   64.07   1982    08  11     NaN   NaN  NaN    NaN
3    1982-08-12   60.52   1982    08  12     NaN   NaN  NaN    NaN
4    1982-08-13   56.87   1982    08  13     NaN   NaN  NaN    NaN
..          ...     ...    ...   ...  ..     ...   ...  ...    ...
969  2022-08-27     NaN   2022    08  27     NaN   NaN  NaN    NaN
970  2022-08-28     NaN   2022    08  28     NaN   NaN  NaN    NaN
971  2022-08-29  900.22   2022    08  29     NaN   NaN  NaN    NaN
972  2022-08-30  890.90   2022    08  30     NaN   NaN  NaN    NaN
973  2022-08-31  882.11   2022    08  31     NaN   NaN  NaN    NaN

[14633 rows x 9 columns]

In [189]:
# Arreglando Dataframes para hacerlos comparables

df_sii_comp = df_sii[df_sii['dia_de_la_semana'] != 'Not valid'].sort_values(by=["Día", "year", "Valor"])
df_sii_comp.reset_index(drop=True, inplace=True)

df_bc_2019_2020_comp = df_bc_2019_2020.sort_values(by=["dia", "year", "Dolar"])
df_bc_2019_2020_comp.reset_index(drop=True, inplace=True)

# Comparando ambos DFs
df_sii_comp[(df_sii_comp["Valor"] != df_bc_2019_2020_comp["Dolar"]) & ~df_sii_comp["Valor"].isnull()]

Empty DataFrame
Columns: [year, mes, Día, Valor, dia_de_la_semana]
Index: []

In [190]:
df_sii_comp.columns = ['dia', 'year', 'mes', 'valor', 'dia_de_la_semana']
df_sii_comp.head()

dia year  mes   valor dia_de_la_semana
0  2019.0  Mar    1  651.79          Viernes
1  2019.0  Feb    1  657.81          Viernes
2  2019.0  Abr    1  678.53            Lunes
3  2019.0  Jul    1  679.15            Lunes
4  2019.0  Ago    1  700.82           Jueves

In [191]:
# Considerando que la transformación de las fechas a días de semana se puede
# realizar utilizando la función get_dow se utilizará eso para df_bc_all

# Sería necesario buscar como extender y hacer merge con otra variable, pero esta solución es más simple

df_bc_all['dia_de_la_semana'] = df_bc_all.apply(lambda x: get_dow(x['dia'], x['mes'], x['year']), axis=1)

display(df_bc_all.head())
display(df_bc_all.tail())


NameError: name 'get_dow' is not defined